bats/tests/helpers/utils.bats

```diff
@@ -0,0 +1,307 @@
+    assert_output foo
+}
+
+@test 'calling_function' {
+    bar() {
+        baz
+    }
+    baz() {
+        calling_function
+    }
+    run bar
+    assert_success
+    assert_output bar
+}
+
+########################################################################
+
+@test 'call_local_function' {
+    local_func() {
+        echo local_func
+    }
+    func() {
+        call_local_function
+    }
+    run func
+    assert_success
+    assert_output local_func
+}
+
+########################################################################
+
+@test 'try will run command at least once' {
+    run try --max 0 --delay 5 inc_counter
+    assert_failure
+    assert_counter_is 1
+    # "try" should not have called "sleep 5" at all
+    ((SECONDS < 2))
+}
+
+@test 'try will stop as soon as the command succeeds' {
+    run try --max 3 --delay 3 inc_counter 2
+    assert_success
+    assert_counter_is 2
+    # "try" should have called "sleep 3" exactly once
+    ((SECONDS >= 3))
+    ((SECONDS < 6))
+}
+
+@test 'try will return after max retries' {
+    run try --max 3 --delay 2 inc_counter
+    assert_failure
+    assert_counter_is 3
+    # "try" should have called "sleep 2" exactly twice
+    ((SECONDS >= 4))
+    ((SECONDS < 6))
+}
+
+########################################################################
+
+@test 'json_string' {
+    run json_string foo\ bar\"baz\'
+    assert_success
+    assert_output "\"foo bar\\\"baz'\""
+}
+
+########################################################################
+
+@test 'join_map echo' {
+    run join_map / echo usr local bin
+    assert_success
+    assert_output usr/local/bin
+}
+
+@test 'join_map false' {
+    run join_map / false usr local bin
+    assert_failure
+}
+
+@test 'join_map json_string' {
+    run join_map ", " json_string true "foo bar" baz:80
+    assert_success
+    assert_output '"true", "foo bar", "baz:80"'
+}
+
+@test 'join_map empty list' {
+    run join_map / echo
+    assert_success
+    assert_output ''
+}
+
+########################################################################
+
+@test 'image_without_tag_as_json_string busybox' {
+    is_quoted busybox
+}
+
+@test 'image_without_tag_as_json_string busybox:latest' {
+    is_quoted busybox
+}
+
+@test 'image_without_tag_as_json_string busybox:5000' {
+    is_quoted busybox
+}
+
+@test 'image_without_tag_as_json_string registry.io:5000' {
+    is_quoted registry.io:5000
+}
+
+@test 'image_without_tag_as_json_string registry.io:5000/busybox' {
+    is_quoted registry.io:5000/busybox
+}
+
+@test 'image_without_tag_as_json_string registry.io:5000/busybox:8080' {
+    is_quoted registry.io:5000/busybox
+}
+
+########################################################################
+
+@test 'unique_filename without extension' {
+    run unique_filename "$COUNTER"
+    assert_success
+    assert_output "${COUNTER}_2"
+    touch "$output"
+
+    run unique_filename "$COUNTER"
+    assert_success
+    assert_output "${COUNTER}_3"
```

Comment:
why dont you want to touch the output on the second time?

Reply:
The only reason to "touch" the output is to make sure the file exists. So when we call `unique_filename` again, it needs to return a different filename.

There is no point in creating the file if we don't call `unique_filename` again afterwards. We are testing `unique_filename`, not our ability to create files.